mountpoint-s3-client/src/mock_client.rs

```diff
@@ -108,13 +108,28 @@ impl MockClient {
             Err(MockClientError("object not found".into()))
         }
     }
+
+    /// Returns error if object does not exist
+    pub fn restore_object(&self, key: &str) -> Result<(), MockClientError> {
+        let mut cloned: MockObject;
+        if let Some(mock_object) = self.objects.write().unwrap().get(key) {
+            cloned = (**mock_object).clone();
+            cloned.restored = true;
+        } else {
+            return Err(MockClientError("object not found".into()));
+        }
+        self.remove_object(key);
+        self.add_object(key, cloned);
+        Ok(())
```

Comment:
Do we need to remove and re-add the object? Maybe I'm missing why it isn't enough to just set `restored` to `true`.

Reply:
So, object, which Arc points to, is immutable. From the [documentation](https://doc.rust-lang.org/std/sync/struct.Arc.html):

> Shared references in Rust disallow mutation by default, and Arc is no exception: you cannot generally obtain a mutable reference to something inside an Arc. If you need to mutate through an Arc, use [Mutex](https://doc.rust-lang.org/std/sync/struct.Mutex.html), [RwLock](https://doc.rust-lang.org/std/sync/struct.RwLock.html), or one of the [Atomic](https://doc.rust-lang.org/std/sync/atomic/index.html) types.

A program like this:
```
use std::{sync::Arc};

struct MockObject {
    restored: bool,
}

fn main() {
    let mut obj1 = Arc::new(MockObject{restored: false});
    obj1.restored = true;
}
```
Won't compile: 
```
error[E0594]: cannot assign to data in an `Arc`
 --> src/main.rs:9:5
  |
9 |     obj1.restored = true;
  |     ^^^^^^^^^^^^^^^^^^^^ cannot assign
  |
  = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `Arc<MockObject>
```

While we can do something like [this](https://doc.rust-lang.org/book/ch16-03-shared-state.html#atomic-reference-counting-with-arct):
```
use std::{sync::{Arc, RwLock}};

struct MockObject {
    restored: bool,
}

fn main() {
    let mut obj1 = Arc::new(RwLock::new(MockObject{restored: false}));
    obj1.write().unwrap().restored = true;
}
```

I am wondering why are we having reference counted objects here in the first place. Probably something simpler will work:
```
use std::{sync::{Arc, RwLock}, collections::BTreeMap};

#[derive(Debug)]
struct MockObject {
    restored: bool,
}

#[derive(Debug)]
struct MockClient {
    objects: RwLock<BTreeMap<String, MockObject>>,  // <-------------------- no Arc's here
}

impl MockClient {
    fn new() -> Self {
        Self {
            objects: RwLock::new(BTreeMap::new()),
        }
    }

    fn add_object(&self, key: &str, value: MockObject) {
        self.objects.write().unwrap().insert(key.to_owned(), value);
    }

    fn restore(&mut self, key: &str) {
        if let Some(mock_object) = self.objects.write().unwrap().get_mut(key) {
            mock_object.restored = true;
        } else {
            panic!("not impl");
        }
    }
}

fn main() {
    let mut client = MockClient::new();
    client.add_object("key", MockObject { restored: false });
    println!("{:?}", client);
    client.restore("key");
    println!("{:?}", client);
}
```

So if you don't mind refactoring this a little, I'll try to remove Arcs tomorrow.

Reply:
Oh right, I missed that it's in an `Arc` (embarrassing since I was the one who did that). It's there because it makes cloning the object cheap, and we make a clone of it when you start a get:

https://github.com/awslabs/mountpoint-s3/blob/b9f7e913fb2e0ddf5beff1f1f3121b4ca1a20601/mountpoint-s3-client/src/mock_client.rs#L318-L323

Without that `Arc`, we'd have to clone the entire object body (the `generator` field in `MockObject`) every time we start a get. BUT! Since we wrote that code, we've moved the `generator` inside `MockObject` into an `Arc` too:

https://github.com/awslabs/mountpoint-s3/blob/b9f7e913fb2e0ddf5beff1f1f3121b4ca1a20601/mountpoint-s3-client/src/mock_client.rs#L114-L120

So `MockObject` should now be fairly cheap to clone even if we get rid of the `Arc<MockObject>` you're tripping over here. Not free to clone — still have to make a copy of the `String`s for `storage_class` and `etag` — but cheap enough. Seems like a good refactor!